src/xdsl/irdl_mlir_printer.py

```diff
@@ -0,0 +1,125 @@
+from xdsl.dialects.irdl import (
+    DialectOp, ParametersOp, TypeOp, OperandsOp, ResultsOp, OperationOp,
+    EqTypeConstraintAttr, AnyTypeConstraintAttr, AnyOfTypeConstraintAttr,
+    DynTypeBaseConstraintAttr, DynTypeParamsConstraintAttr,
+    TypeParamsConstraintAttr, NamedTypeConstraintAttr)
+from dataclasses import dataclass
+
+
+@dataclass(frozen=True, eq=False)
+class IRDLPrinter:
+
+    stream: IOBase
+
+    def _print(self, s: str, **kw_args):
+        print(s, file=self.stream, **kw_args)
+
+    def print_module(self, module):
+        self._print('module{')
+        module.walk(lambda di: IRDLPrinter.print_dialect_definition(self, di)
+                    if isinstance(di, DialectOp) else None)
+        self._print('}')
+
+    def print_type_definition(self, type: TypeOp):
+        self._print(f"    {type.name} {type.attributes['type_name'].data} {{")
+        type.walk(lambda param: self.print_parameters_definition(param)
+                  if isinstance(param, ParametersOp) else None)
+        self._print(f"    }}")
+
+    def print_attr_constraint(self, f):
+
+        if isinstance(f, AnyOfTypeConstraintAttr):
+            self._print(f"AnyOf", end="<")
+            for i in range(len(f.params.data) - 1):
+                self.print_attr_constraint(f.params.data[i])
+                self._print(", ", end="")
+            self.print_attr_constraint(f.params.data[len(f.params.data) - 1])
+            self._print(">", end="")
+
+        elif isinstance(f, EqTypeConstraintAttr):
+            p = Printer(self.stream)
+            p.print(f.type)
+
+        elif isinstance(f, AnyTypeConstraintAttr):
+            self._print(f"Any", end="")
+
+        elif isinstance(f, DynTypeBaseConstraintAttr):
+            self._print(f.type_name.data, end='')
+
+        elif isinstance(f, DynTypeParamsConstraintAttr):
+            self._print(f.type_name.data, end='')
+            self._print("<", end="")
+            for i in range(len(f.params.data)):
+                self.print_attr_constraint(f.params.data[i])
+            self._print(">", end="")
+
+        elif isinstance(f, TypeParamsConstraintAttr):
+            self._print(f"{f.type_name}")
+
+        elif isinstance(f, NamedTypeConstraintAttr):
+            self._print(f.type_name.data, end='')
+            self._print(":", end=" ")
+            self.print_attr_constraint(f.params_constraints)
+
+        else:
+            raise Exception(f"Invalid Constraint: {type(f)} ")
+
+    def print_parameters_definition(self, param_op: ParametersOp):
+        self._print(f"      {ParametersOp.name}", end="(")
+
+        for param in param_op.constraints.data:
+            self._print(f"{param.type_name.data}: ", end="")
+            self.print_attr_constraint(param.params_constraints)
+        self._print(")")
+
+    def print_operation_definition(self, operation: OperationOp):
+        self._print(
+            f"    {OperationOp.name} {operation.attributes['name'].data} {{")
+        '''
```

Comment:
Comments should use `#` rather than the inline strings.
We use only the `'''` for the documentation comments, which are directly after a function or class definition.